### PR TITLE
feat: bump up harbor to v2.2.1

### PR DESCRIPTION
**Pull request by heww**
Signed-off-by: He Weiwei <hweiwei@vmware.com>

### CODE REVIEW

**Review by bitsf — APPROVED** [review]

**Comment by steven-zou** [comment]
@holyhope @sguyennet 

Would you please take a look at this PR?  Thanks！

File: controllers/goharbor/core/deployments.go [hunk 1]
```diff
@@ -30,34 +30,54 @@ var (
 	fsGroup    int64 = 10000
 	runAsGroup int64 = 10000
 	runAsUser  int64 = 10000
+
+	metricNamespace = "harbor"
+	metricSubsystem = "core"
```
Comment (holyhope) [inline controllers/goharbor/core/deployments.go:35]: Are these constant not accessible from harbor repository?
I suggest to import them if possible so next update will be easier using go get -u`

File: controllers/goharbor/core/deployments.go [hunk 2]
```diff
@@ -30,34 +30,54 @@ var (
...
+func getDefaultAllowedRegistryTypesForProxyCache() string {
+	return strings.Join([]string{
+		DockerHubRegistryTypesForProxyCache,
+		HarborRegistryTypesForProxyCache,
+		ACRRegistryTypesForProxyCache,
+		ECRRegistryTypesForProxyCache,
+		GCRRegistryTypesForProxyCache,
+		QuayRegistryTypesForProxyCache,
+	}, ",")
+}
```
Comment (holyhope) [inline controllers/goharbor/core/deployments.go:79]: Sam suggestions here: use constants from harbor repository if possible

File: controllers/goharbor/core/deployments.go [hunk 3]
```diff
@@ -200,6 +220,13 @@ func (r *Reconciler) GetDeployment(ctx context.Context, core *goharborv1.Core) (
 		return nil, errors.Wrap(err, "cannot configure environment variables")
 	}
 
+	metricsEnvs, err := r.getMetricsEnvVars(core)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot configure environment variables")
```
Comment (holyhope) [inline controllers/goharbor/core/deployments.go:225]:
```suggestion
		return nil, errors.Wrap(err, "get metrics environment variables")
```

File: controllers/goharbor/harbor/resources.go [hunk 4]
```diff
@@ -90,6 +90,16 @@ func (r *Reconciler) AddResources(ctx context.Context, resource resources.Resour
 		return errors.Wrapf(err, "cannot add %s", controllers.ChartMuseum)
 	}
 
+	exporterCertificate, err := r.AddExporterConfigurations(ctx, harbor, internalTLSIssuer)
+	if err != nil {
+		return errors.Wrapf(err, "cannot add %s configuration", controllers.Exporter)
```
Comment (holyhope) [inline controllers/goharbor/harbor/resources.go:95]: I suggest to no more use sentences in error wrapping.
On new feature that I pushed, I am adding only context without the _negative_ word.
Example:

```suggestion
		return errors.Wrapf(err, "add %s configuration", controllers.Exporter)
```

This way, errors serialisation is shorter.
WDYT?

PS: I know that I was the one who wrote it first, but I now think that I was wrong ☺️

File: pkg/config/harbor/env.go [hunk 5]
```diff
@@ -41,18 +41,18 @@ func EnvVar(configName string, value ConfigValue) (corev1.EnvVar, error) {
 }
 
 func EnvVars(configs map[string]ConfigValue) ([]corev1.EnvVar, error) {
-	errors := []corev1.EnvVar{}
+	envVars := []corev1.EnvVar{}
```
Comment (holyhope) [inline pkg/config/harbor/env.go:44]: 👍😁

File: pkg/scheme/scheme_test.go [hunk 6]
```diff
@@ -37,6 +37,10 @@ var _ = Describe("Runtime scheme ", func() {
 			Group:   "extensions",
 			Version: "v1beta1",
 			Kind:    "Ingress",
+		}), Entry("v1alpha3.goharbor.io/Habor", schema.GroupVersionKind{
```
Comment (holyhope) [inline pkg/scheme/scheme_test.go:40]:
```suggestion
		}), Entry("v1alpha3.goharbor.io/Harbor", schema.GroupVersionKind{
```

File: pkg/scheme/scheme_test.go [hunk 7]
```diff
@@ -37,6 +37,10 @@ var _ = Describe("Runtime scheme ", func() {
...
+			Group:   "goharbor.io",
+			Version: "v1alpha3",
+			Kind:    "Exporter",
 		}), Entry("v1alpha3.goharbor.io/Habor", schema.GroupVersionKind{
```
Comment (holyhope) [inline pkg/scheme/scheme_test.go:44]:
```suggestion
 		}), Entry("v1alpha3.goharbor.io/Harbor", schema.GroupVersionKind{
```

**Review by holyhope — CHANGES_REQUESTED** [review]
Can you please `go get github.com/goharbor/harbor@v2.2.1` please?
So we keep all constants and config up to date.

File: controllers/goharbor/harbor/resources.go (hunk 4 above)
Comment (steven-zou) [reply controllers/goharbor/harbor/resources.go:95]: Actually, that's a better way for an error statement. I just have one minor concern, should that end with an `error` word?

```go
return errors.Wrapf(err, "add %s configuration error", controllers.Exporter)
```

**Comment by heww** [comment]
> Can you please `go get github.com/goharbor/harbor@v2.2.1` please?
> So we keep all constants and config up to date.

Due to the limitation of the harbor repo (`go.mod` and `go.sum` are in the src subdirectory not in the root directory of the repo), `go get github.com/goharbor/harbor@v2.2.1` not works as we think.

Update the `github.com/goharbor/harbor/src` also not works.
``` 
go get github.com/goharbor/harbor/src@v2.2.1
go get github.com/goharbor/harbor/src@v2.2.1: github.com/goharbor/harbor/src@v2.2.1: invalid version: unknown revision src/v2.2.1
```

So I will just update the `github.com/goharbor/harbor/src`

File: controllers/goharbor/core/deployments.go (hunk 1 above)
Comment (heww) [reply controllers/goharbor/core/deployments.go:35]: These constants are not accessible from the harbor repository.

File: controllers/goharbor/harbor/resources.go (hunk 4 above)
Comment (holyhope) [reply controllers/goharbor/harbor/resources.go]: Not sure about the `error` keyword.
The value is already `error` typed.
the `wrap()` feature concatenate strings, so it is not necessary to have `error`:
1. ```
   add resources: add exporter configuration: get certificate
   ```
2. ```
   add resources error: add exporter configuration error: get certificate error
   ```
IMO 1 is more readable.

File: controllers/goharbor/exporter/deployments.go [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+	} else if scheme == "https" {
+		port = "443"
+	}
+
+	parts := strings.Split(u.Host, ":")
+	host = parts[0]
+
+	if len(parts) == 2 { // nolint:gomnd
+		port = parts[1]
+	}
```
Comment (holyhope) [inline controllers/goharbor/exporter/deployments.go:238]: I suggest something that imo is more readable (no split and array index):

```suggestion
	scheme = u.Scheme
	host = u.Hostname()
	port = u.Port()
	
	if port == "" {
		switch strings.ToLower(scheme) {
			case "http":
				port = "80"
			default:
				port = "443"
		}
	}
```

File: controllers/goharbor/exporter/networkpolicies.go [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+type NetworkPolicy graph.Resource
+
+func (r *Reconciler) AddNetworkPolicies(ctx context.Context, exporter *goharborv1.Exporter) error {
+	areNetworkPoliciesEnabled, err := r.AreNetworkPoliciesEnabled(ctx, exporter)
+	if err != nil {
+		return errors.Wrapf(err, "cannot get status")
```
Comment (holyhope) [inline controllers/goharbor/exporter/networkpolicies.go:20]: I suppose that you copy-pasted it from somewhere else, but I suggest to change this error message which does not match the previous call.

```suggestion
		return errors.Wrapf(err, "check network policy")
```

Pease tell me if it is out of the scope, it wan be ^^'

File: controllers/goharbor/harbor/exporter.go [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+	namespace := harbor.GetNamespace()
+
+	tls := harbor.Spec.InternalTLS.GetComponentTLSSpec(r.GetInternalTLSCertificateSecretName(ctx, harbor, harbormetav1.ExporterTLS))
+
+	host := r.NormalizeName(ctx, harbor.GetName(), controllers.Core.String())
+	if harbor.Spec.InternalTLS.IsEnabled() {
+		host += ":443"
+	} else {
+		host += ":80"
+	}
```
Comment (holyhope) [inline controllers/goharbor/harbor/exporter.go:73]: Is it really useful? Scheme is already specified line 76.

```suggestion
```

File: controllers/goharbor/harbor/resources.go [hunk 11]
```diff
@@ -22,102 +22,112 @@ func (r *Reconciler) AddResources(ctx context.Context, resource resources.Resour
 
 	_, _, internalTLSIssuer, err := r.AddInternalTLSConfiguration(ctx, harbor)
 	if err != nil {
-		return errors.Wrap(err, "cannot add internal TLS configuration")
+		return errors.Wrap(err, "add internal TLS configuration error")
```
Comment (holyhope) [inline controllers/goharbor/harbor/resources.go:25]:
```suggestion
		return errors.Wrap(err, "add internal TLS configuration")
```

File: controllers/goharbor/harbor/resources.go [hunk 12]
```diff
@@ -22,102 +22,112 @@ func (r *Reconciler) AddResources(ctx context.Context, resource resources.Resour
...
 	}
 
 	registryCertificate, registryAuthSecret, registryHTTPSecret, err := r.AddRegistryConfigurations(ctx, harbor, internalTLSIssuer)
 	if err != nil {
-		return errors.Wrapf(err, "cannot add %s configuration", controllers.Registry)
+		return errors.Wrapf(err, "add %s configuration error", controllers.Registry)
```
Comment (holyhope) [inline controllers/goharbor/harbor/resources.go:30]:
```suggestion
		return errors.Wrapf(err, "add %s configuration", controllers.Registry)
```

**Review by holyhope — CHANGES_REQUESTED** [review]

File: controllers/goharbor/harbor/exporter.go (hunk 10 above)
Comment (heww) [reply controllers/goharbor/harbor/exporter.go:73]: This is the same logic to get the core URL in the jobservice. I will extract these lines to be a method, jobservice and exporter will use this new method to get the URL of the core.

File: controllers/goharbor/exporter/networkpolicies.go (hunk 9 above)
Comment (heww) [reply controllers/goharbor/exporter/networkpolicies.go:20]: Let's keep `return errors.Wrapf(err, "cannot get status")` same with other components.

If really need to change this error message, I will open a new PR to change this message in the whole source code.

File: apis/meta/v1alpha1/metric.go [hunk 13]
```diff
@@ -0,0 +1,45 @@
+package v1alpha1
+
+import (
+	"fmt"
+)
+
+const (
+	PrometheusScrapeAnnotationKey = "prometheus.io/scrape"
+	PrometheusPathAnnotationKey   = "prometheus.io/path"
+	PrometheusPortAnnotationKey   = "prometheus.io/port"
```
[the remark this answers is not in the document]
Comment (steven-zou) [reply apis/meta/v1alpha1/metric.go:10]: @holyhope let's create an issue to track the next action.

File: controllers/goharbor/core/deployments.go [hunk 14]
```diff
@@ -30,34 +31,45 @@ var (
...
+func getDefaultAllowedRegistryTypesForProxyCache() string {
+	return strings.Join([]string{
+		registry.RegistryTypeDockerHub,
+		registry.RegistryTypeHarbor,
+		registry.RegistryTypeAzureAcr,
+		registry.RegistryTypeAwsEcr,
+		registry.RegistryTypeGoogleGcr,
+		registry.RegistryTypeQuay,
+	}, ",")
+}
```
[the remark this answers is not in the document]
Comment (heww) [reply controllers/goharbor/core/deployments.go:71]: This setting is not configurable for the customers so I don't think we need to move it to `/apis/...` like we get the TrivyVulnerabilityType.

**Review by holyhope — APPROVED** [review]

**Comment by holyhope** [comment]
LGTM :-)

**Pull request merged**